README.md
ffuzzy++ : C++ implementation of ssdeep-compatible fast fuzzy hashing
======================================================================



What is ffuzzy++?
------------------

ffuzzy++ is a fuzzy hash implementation compatible with
ssdeep version 2.13 (<http://ssdeep.sourceforge.net/>).
Using this implementation, you can:

*	Generate fuzzy digests from files and buffers
*	Compare fuzzy digests generated from ffuzzy++ or
	other ssdeep-compatible implementations

This implementation is designed to be fast, stable and thread-safe.
Sometimes this is about 60% to 120% faster than ssdeep when you
cluster digests and is about 10% to 60% faster than ssdeep when
you generate digests.

However, this C++ implementation is primarily designed to be fast
by exposing internal structure of fuzzy hashing. So this
implementation may not be safe as you think (depends on function
you use). Instead, you can boost your clustering application
using efficient digest utilities.



Requirements
-------------

You will need C++ compiler with full C++11 support.

You will also need GNU Autoconf Archive when you build `configure`
script (<https://www.gnu.org/software/autoconf-archive/>).

If you wish to run the test, Google C++ Testing Framework
(<https://code.google.com/p/googletest/>) will be needed.



Installation
-------------

This program can be installed without proper "installation".
Just copy `ffuzzy.hpp` and `ffuzzypp` directory to anywhere you want.
In this case, you need to define `FFUZZYPP_DECLARATIONS` before 
`ffuzzy.hpp` in **one** of the source files to make the linker happy.

You can compile and install a static helper library `libffuzzypp.a`
by running `./configure && make && make install`.
If you don't have `configure` script on the top directory, you need
to run `./bootstrap.sh` to make build files.



Configuration Options
----------------------

*	--disable-static-lib  
	Disable building `libffuzzypp.a`
*	--disable-debug  
	Disable assertions on examples and tests.
	Don't disable assertions if you wish to test ffuzzy++.
*	--enable-examples  
	Enable building examples (see "Building Examples" section)
*	--enable-tests  
	Enable building test cases (see "Running Tests" section)
*	--disable-compatibility-tests  
	Ignore ssdeep-specific parameters during compilation of
	test cases. This option prevents compiling compatibility tests.



Building Examples
------------------

There are two examples:

*	examples/compare-hash  
	You can use this program to compare two fuzzy digests.
*	examples/compute-hash  
	You can use this program to compute fuzzy digests from
	given files.



Running Tests
--------------

By default, compatibility tests (which depends on ssdeep-specific
parameters) are enabled and you can disable these tests by adding
`--disable-compatibility-tests` configuration option.

*	tests/test-small  
	Relatively small tests (that can be done in a few minutes).
*	tests/test-compatibility-small  
	Compatibility tests (depends on ssdeep parameters)
*	tests/test-compatibility-small-2.9  
	Compatibility tests to test ssdeep 2.9 emulation
*	tests/test-compatibility-large  
	Large compatibility tests (depends on ssdeep parameters)
*	tests/test-precond  
	Precondition tests (it always succeeds at runtime because
	all "tests" are performed while compiling this executable).



Caution
--------

*	This package is intended for static-linking to your program.
*	Packaging this implementation **alone** for OS distribution
	is not recommended.  The interface may change without notice.
*	Digests generated from files equal to or larger than 4GiB
	are not portable enough. See following secions for details.



Differences between ssdeep
---------------------------

When comparing identical digests, ssdeep 2.9 and 2.13 return different
values. By default, ffuzzy++ emulates behavior of version 2.13.
If you define `FFUZZYPP_COMPATIBILITY_SSDEEP_2_9`, ffuzzy++ will
emulate behavior of version 2.9.

Result of comparison is not in a signed type but in an unsigned type
because error handling things are moved out of comparison.
So comparison functions cannot return -1 on error (like ssdeep/libfuzzy).
Instead, digest parser (which will be used before comparison) will raise
an exception when it fails to parse digest string.



Differences between ssdeep -2.12 (DETAILS)
-------------------------------------------

Because of current implementation restrictions in ssdeep 2.12 and before
(not in ffuzzy++), the digest and comparison result may different when:

*	Extremely large files are given (for example):
	*	For files equal to or larger than 4GiB,  
		ssdeep 2.10-2.12 on 32-bit platform and ssdeep -2.9
		causes arithmetic overflow while handling file sizes.
		This will result in unexpected behavior.
	*	For files larger than 4GiB,  
		ssdeep -2.12's rolling hash implementation is not exactly
		rolling. This will result in incorrect digest value.
	*	For files larger than 96GiB,  
		ssdeep 2.10-2.12 will crash under some circumstances.
		ssdeep 2.9 with 64-bit file size patch will use incorrect value
		(because of arithmetic overflow) when generating digests.  
		ffuzzy++'s implementation matches mathematical properties
		of ssdeep digest generation algorithm while preventing
		arithmetic overflow bug on ssdeep 2.9 with
		64-bit file size patch.
	*	For files equal to or larger than 2^64 bytes,  
		ssdeep 2.10-2.12 on 64-bit platform will cause arithmetic
		overflow on file sizes. Although we don't accept
		files larger than 192GiB (see "Limitations" section), ssdeep
		-2.12 doesn't have mechanism to prevent file size overflow.
*	Some digests generated from extremely large files are given:
	*	For digests generated from files larger than 12GiB,  
		ssdeep -2.12 may cause arithmetic overflow while comparing
		digests. This may make incorrect comparison score of zero.

Overrall, ffuzzy++ is almost compatible with ssdeep -2.12. But on some
circumstances which ssdeep will fail or did not support (yes, actually
files equal to or larger than 4GiB are not officially supported by
ssdeep -2.12), ffuzzy++ returns mostly legit values
(based on mathematical properties of ssdeep).



Differences between ssdeep 2.13
--------------------------------

ssdeep 2.13 behaves very similar to ffuzzy++ due to bug fixes.
There are few implementation differences but behaves the same when:

*	Generating fuzzy digests from arbitrary input
*	Comparing digests generated by new implementations
*	Comparing digests generated from files up to 96GiB
	by previous ssdeep versions



Limitations
------------

Both implementations (ssdeep and ffuzzy++) have a limit of 192GiB in
digest generator. Based on mathematical properties of ssdeep,
we **could** accept much larger files. However, digests generated
from such files are nearly useless (because resulting digests will
not represent signature of whole file contents).

So I decided not to accept files larger than 192GiB on ffuzzy++
as well as ssdeep 2.10+.

Although ffuzzy++ implementation is completely platform-independent,
ssdeep digest for files equal to or larger than 4GiB are still not
portable enough (varies by ssdeep versions and platforms).

Make sure that you use new implementations (ssdeep 2.13 or ffuzzy++)
when you accept extremely large files. If you use older versions of
ssdeep, please consider rejecting such big files (equal to or larger
than 4GiB) when you generate digests.



License / Copying
------------------

See `COPYING.md` file for details.


NEWS.md
ffuzzy++ : C++ implementation of fast fuzzy hashing
====================================================


Version 4.0.1 - 2017-06-28
---------------------------

*	`make dist` issue is now fixed.


Version 4.0 - 2017-06-28
-------------------------

*	NEW: Bit-parallelism  
	(clustering can be over 10 times faster than version 3.0)
*	Digest generator is optimized  
	(can be twice as fast as version 3.0)
*	Utility methods are added


Version 3.0.1 - 2017-06-04
---------------------------

*	Uses of deprecates type trait are now replaced with decent ones.


Version 3.0 - 2015-04-25
-------------------------

*	The initial public release


ffuzzypp/base64.hpp
/*

	ffuzzy++ : C++ implementation of fast fuzzy hashing

	base64.hpp
	Base64 utilities

	Copyright (C) 2017 Tsukasa OI <floss_ssdeep@irq.a4lg.com>


	Permission to use, copy, modify, and/or distribute this software for
	any purpose with or without fee is hereby granted, provided that the
	above copyright notice and this permission notice appear in all copies.

	THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
	WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
	MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
	ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
	WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
	ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
	OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

*/
#ifndef FFUZZYPP_BASE64_HPP
#define FFUZZYPP_BASE64_HPP

namespace ffuzzy {

class base64
{
private:
	base64(void) = delete;
	base64(const base64&) = delete;

	// Base64 table
public:
	static constexpr const char invalid_index = 64;
	static constexpr const char values[64] =
	{
		// This should be the same order of ordinal Base64
		// (as well as b64 variable in fuzzy.c)
		'A', 'B', 'C', 'D', 'E', 'F', 'G',
		'H', 'I', 'J', 'K', 'L', 'M', 'N',
		'O', 'P', 'Q', 'R', 'S', 'T', 'U',
		'V', 'W', 'X', 'Y', 'Z',
		'a', 'b', 'c', 'd', 'e', 'f', 'g',
		'h', 'i', 'j', 'k', 'l', 'm', 'n',
		'o', 'p', 'q', 'r', 's', 't', 'u',
		'v', 'w', 'x', 'y', 'z',
		'0', '1', '2', '3', '4',
		'5', '6', '7', '8', '9',
		'+', '/'
	};
	static_assert(
		sizeof(values) == 64 &&
		values[ 0] == 'A' &&
		values[26] == 'a' &&
		values[52] == '0' &&
		values[62] == '+' &&
		values[63] == '/',
		"minimum sanity check for Base64 failed.");
	static_assert(
		invalid_index < 0 || invalid_index >= 64,
		"invalid_index is not out of range.");

public:
	// Utility to check whether given character is in Base64 charset
	static bool isbase64(char c) noexcept
	{
		switch (c)
		{
		case 'A': case 'B': case 'C': case 'D': case 'E': case 'F': case 'G':
		case 'H': case 'I': case 'J': case 'K': case 'L': case 'M': case 'N':
		case 'O': case 'P': case 'Q': case 'R': case 'S': case 'T': case 'U':
		case 'V': case 'W': case 'X': case 'Y': case 'Z':
		case 'a': case 'b': case 'c': case 'd': case 'e': case 'f': case 'g':
		case 'h': case 'i': case 'j': case 'k': case 'l': case 'm': case 'n':
		case 'o': case 'p': case 'q': case 'r': case 's': case 't': case 'u':
		case 'v': case 'w': case 'x': case 'y': case 'z':
		case '0': case '1': case '2': case '3': case '4':
		case '5': case '6': case '7': case '8': case '9':
		case '+': case '/':
			return true;
		default:
			return false;
		}
	}
	// Utility to convert character to Base64 index
	static char toindex(char c) noexcept
	{
		switch (c)
		{
		case 'A': return  0;
		case 'B': return  1;
		case 'C': return  2;
		case 'D': return  3;
		case 'E': return  4;
		case 'F': return  5;
		case 'G': return  6;
		case 'H': return  7;
		case 'I': return  8;
		case 'J': return  9;
		case 'K': return 10;
		case 'L': return 11;
		case 'M': return 12;
		case 'N': return 13;
		case 'O': return 14;
		case 'P': return 15;
		case 'Q': return 16;
		case 'R': return 17;
		case 'S': return 18;
		case 'T': return 19;
		case 'U': return 20;
		case 'V': return 21;
		case 'W': return 22;
		case 'X': return 23;
		case 'Y': return 24;
		case 'Z': return 25;
		case 'a': return 26;
		case 'b': return 27;
		case 'c': return 28;
		case 'd': return 29;
		case 'e': return 30;
		case 'f': return 31;
		case 'g': return 32;
		case 'h': return 33;
		case 'i': return 34;
		case 'j': return 35;
		case 'k': return 36;
		case 'l': return 37;
		case 'm': return 38;
		case 'n': return 39;
		case 'o': return 40;
		case 'p': return 41;
		case 'q': return 42;
		case 'r': return 43;
		case 's': return 44;
		case 't': return 45;
		case 'u': return 46;
		case 'v': return 47;
		case 'w': return 48;
		case 'x': return 49;
		case 'y': return 50;
		case 'z': return 51;
		case '0': return 52;
		case '1': return 53;
		case '2': return 54;
		case '3': return 55;
		case '4': return 56;
		case '5': return 57;
		case '6': return 58;
		case '7': return 59;
		case '8': return 60;
		case '9': return 61;
		case '+': return 62;
		case '/': return 63;
		default:  return invalid_index;
		}
	}

	// Base64 transformation (from index to actual character)
public:
	class transform_to_b64
	{
	private:
		transform_to_b64(void) = delete;
		transform_to_b64(const transform_to_b64&) = delete;
	public:
		static constexpr char transform(char ch) noexcept
		{
			return base64::values[static_cast<unsigned char>(ch)];
		}
	};

	// Base64 reverse-transformation (from Base64 character to index)
public:
	class transform_from_b64
	{
	private:
		transform_from_b64(void) = delete;
		transform_from_b64(const transform_from_b64&) = delete;
	public:
		static char transform(char ch) noexcept
		{
			return toindex(ch);
		}
	};
};

#ifdef FFUZZYPP_DECLARATIONS
constexpr const char base64::values[64];
#endif

}

#endif
